### PR TITLE
feat!: route implementation

### LINKED ISSUE #8: Support for Webhooks

### PR DESCRIPTION
**Pull request by VincentRPS**
## Summary

The biggest thing currently lacking on v3 is route support.
This initiates some new routes for users to play around with.

Closes #8, #7, and #6.

## PR Type

- [x] This PR improves the code (speed/style improvements or refactoring).
- [x] This PR fixes an issue.
- [x] This PR makes an addition to the library (new file, class, method, parameters, etc.).
- [x] This PR makes a breaking change (renaming/removing parameters, methods, etc.).
- [ ] This PR doesn't make any code changes to the library (documentation, examples).

## Checklist

- [x] I have searched all pull requests for duplicates.
- [x] I have tested the changes made.
  - [ ] I have properly changed the docs to reflect the changes made.
- [x] If 'type: ignore' comments were used, a comment is left to explain why it is needed.

### CODE REVIEW

**Comment by VincentRPS** [comment]
@plun1331 what if for the `roles` and `channels` parameters on `Bot.create_guild`, we created a new class called `ConstructableRole` and `ConstructableChannel` which is what the user would use for it?

**Comment by plun1331** [comment]
> @plun1331 what if for the `roles` and `channels` parameters on `Bot.create_guild`, we created a new class called `ConstructableRole` and `ConstructableChannel` which is what the user would use for it?

Sounds good to me

File: pycord/pages/paginator.py [hunk 1]
```diff
@@ -18,12 +18,12 @@
...
-from typing import Protocol, TypeVar
+from typing import Any, Generic, Protocol, TypeVar
 
 from .errors import NoMorePages, PagerException
 
 T = TypeVar('T', covariant=True)
-
+P = TypeVar('P', bound='Page')
 
 __all__ = ['Page', 'Paginator']
```
Comment (ooliver1) [inline pycord/pages/paginator.py:28]: `__all__` does not need to be dynamic, this is a waste of memory and `__all__` should be immutable.

File: pycord/pages/paginator.py [hunk 2]
```diff
@@ -33,8 +33,7 @@ class Page(Protocol[T]):
     The class for all Page Types to subclass.
     """
 
-    def __init__(self):
-        ...
+    value: Any
```
Comment (ooliver1) [inline pycord/pages/paginator.py:36]: Then what's the point of `T`? Should it not use `T`?

File: pycord/snowflake.py [hunk 3]
```diff
@@ -46,3 +46,7 @@ def increment(self) -> int:
 
     def __hash__(self) -> int:
         return self >> 22
+
+    @classmethod
+    def from_datetime(cls, dt: datetime) -> 'Snowflake':
```
Comment (ooliver1) [inline pycord/snowflake.py:51]: Or alternatively use `from __future__ import annotations` and remove the quotes.

File: pycord/typing.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import typing
+from typing import TYPE_CHECKING, Any
+
+from .types.snowflake import Snowflake
+
+if TYPE_CHECKING:
+    from .state import State
+
+
+class Typing:
```
Comment (ooliver1) [inline pycord/typing.py:32]:
```suggestion
class Typing:
    __slots__ = ('_state', '__channel_id')

```

File: pycord/types/application_role_connection_metadata.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    4,
+    5,
+    6,
+    7,
+    8,
+]
+
+
+class ApplicationRoleConnectionMetadata(TypedDict):
+    type: RCMTYPE
```
[the remark this answers is not in the document]
Comment (VincentRPS) [reply pycord/types/application_role_connection_metadata.py:41]: It's a base constant (as in, it's a non-changing variable in the file) so by standards and by other modules in .types, it should be fully capitalized

File: pycord/channel.py [hunk 6]
```diff
@@ -122,7 +251,7 @@ def __init__(self, data: DiscordChannel, state: State) -> None:
         )
         self.position: int | UndefinedType = data.get('position', UNDEFINED)
         self.permission_overwrites: list[_Overwrite] = [
-            _Overwrite(d) for d in data.get('permission_overwrites', [])
+            _Overwrite.from_dict(d) for d in data.get('permission_overwrites', [])
         ]
         self.topic: str | None | UndefinedType = data.get('topic', UNDEFINED)
```
[the remark this answers is not in the document]
Comment (VincentRPS) [reply pycord/channel.py:256]: Well the intended functionality is if a user does something like `if channel.topic` not `if channel.topic is None`

File: pycord/channel.py [hunk 7]
```diff
@@ -161,17 +290,16 @@ def __init__(self, data: DiscordChannel, state: State) -> None:
...
+        *,
         nonce: int | str | UndefinedType = UNDEFINED,
         tts: bool | UndefinedType = UNDEFINED,
         embeds: list[Embed] | UndefinedType = UNDEFINED,
+        allowed_mentions: AllowedMentions | UndefinedType = UNDEFINED,
+        message_reference: MessageReference | UndefinedType = UNDEFINED,
         sticker_ids: list[Snowflake] | UndefinedType = UNDEFINED,
         flags: int | UndefinedType = UNDEFINED,
-        house: House | UndefinedType = UNDEFINED,
         houses: list[House] | UndefinedType = UNDEFINED,
```
[the remark this answers is not in the document]
Comment (VincentRPS) [reply pycord/channel.py:301]: The House API is going to be removed and get a complete facelift once this PR is done. Components are a really big priority for me, but actual functionality (what this PR adds) is more.

File: pycord/flags.py [hunk 8]
```diff
@@ -550,3 +550,22 @@ def bot_http_interactions(self) -> bool | int:
...
+
+
+@fill()
+class MemberFlags(Flags):
+    @flag
+    def did_rejoin(self) -> bool | int:
+        return 1 << 0
+
+    @flag
+    def completed_onboarding(self) -> bool | int:
```
[the remark this answers is not in the document]
Comment (VincentRPS) [reply pycord/flags.py:562]: If you access it via `MemberFlags.completed_onboarding`, it's an integer value. But on something like `MemberFlags().completed_onboarding` it's a book.

File: pycord/guild.py [hunk 9]
```diff
@@ -47,12 +71,37 @@
...
+        *,
+        lock_permissions: bool = False,
+        parent_id: Snowflake | None | UndefinedType,
+    ) -> None:
+        self.id = id
+        self.position = position
+        self.lock_permissions = lock_permissions
+        self.parent_id = parent_id
+
+    def to_dict(self) -> dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (VincentRPS) [reply pycord/guild.py:101]: there doesn't seem to be a type for it in `pycord.types`, so I'll place this into a TODO

File: pycord/message.py [hunk 10]
```diff
@@ -100,6 +104,39 @@ def __init__(self, data: DiscordMessageReference) -> None:
...
 
+    # refactor so user can initialize
+    def __init__(
+        self,
+        message_id: Snowflake | UndefinedType = UNDEFINED,
+        channel_id: Snowflake | UndefinedType = UNDEFINED,
+        guild_id: Snowflake | UndefinedType = UNDEFINED,
+        fail_if_not_exists: bool | UndefinedType = UNDEFINED,
+    ) -> None:
+        self.message_id: Snowflake | UndefinedType = message_id
```
[the remark this answers is not in the document]
Comment (VincentRPS) [reply pycord/message.py:123]: It is, but it just felt better using UNDEFINED here most likely.

File: pycord/channel.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ooliver1) [reply pycord/channel.py:307]: But why can it be **two** different sentinels without any semantic meaning? This is the entire use of `None`. The only point of `UNDEFINED` is to know if a parameter has not been passed, not to abuse in debt of the user.

File: pycord/flags.py [hunk 11]
```diff
@@ -550,3 +550,22 @@ def bot_http_interactions(self) -> bool | int:
     @flag
     def active_developer(self) -> bool | int:
         return 1 << 22
+
+
+@fill()
+class MemberFlags(Flags):
+    @flag
+    def did_rejoin(self) -> bool | int:
```
[the remark this answers is not in the document]
Comment (VincentRPS) [reply pycord/flags.py:558]: For something as small as flags, I doubt documentation is really needed. The name explains it all, and having to document hundreds of flags is something no one wants to do.

**Comment by VincentRPS** [comment]
I will be giving some time for people to review.

Will be merging near 7PM (UTC+8) if all is resolved or no more reviews are given.

**Comment by plun1331** [comment]
> ## Checklist
> * [x]  I have searched all pull requests for duplicates.
> * [ ]  I have tested the changes made.
>   
>   * [ ]  I have properly changed the docs to reflect the changes made.
> * [x]  If 'type: ignore' comments were used, a comment is left to explain why it is needed.

did you test the changes

**Pull request merged**